### PR TITLE
Enable RotatedSunFrame to work with non-SunPy frames

### PR DESCRIPTION
**Pull request by kathivashkiv**
Edit by @ayshih : This PR enables `RotatedSunFrame` to work with non-SunPy frames.  This change required our minimum Astropy version to be at least 4.0, so this PR should not be backported.

There are two changes for #4000 issue

### CODE REVIEW

**Comment by ayshih** [comment]
This PR is actually a feature PR.  I have changed the title/descriptions/labels accordingly.  Please:

- Add a changelog entry (`changelog/4577.feature.rst`)
- Add a test (`sunpy/coordinates/tests/test_metaframes.py`) using a non-SunPy frame (e.g., `HeliocentricMeanEcliptic`).  The test doesn't need to be as extensive as all of the other tests in that file.

**Comment by kathivashkiv** [comment]
@ayshih there is one test that fails after changes (in sunpy/coordinates/tests/test_metaframes.py): test_no_sunpy_frame()
Should I delete it or rewrite this test for new behavior?

**Comment by ayshih** [comment]
Yeah, you'd replace that test.  Basically, test that `RotatedSunFrame(base=HeliocentricMeanEcliptic(obstime=...))` does not raise an error.

File: sunpy/coordinates/tests/test_metaframes.py
Comment (ayshih) [inline sunpy/coordinates/tests/test_metaframes.py:118]:
```suggestion
def test_non_sunpy_frame():
```

File: changelog/4577.feature.rst
Comment (ayshih) [inline changelog/4577.feature.rst:1]:
```suggestion
Enable `~sunpy.coordinates.metaframes.RotatedSunFrame` to work with non-SunPy frames (e.g., `~astropy.coordinates.HeliocentricMeanEcliptic`).
```

**Review by ayshih — CHANGES_REQUESTED** [review]
Please add a second changelog entry (`4577.breaking.rst`) saying, "The ``.size`` property of a coordinate frame with no associated data will now raise an error instead of returning 0."

**Comment by kathivashkiv** [comment]
Is this a specific error (e.g., ValueError), or saying "will now raise an error" is enough?

**Comment by ayshih** [comment]
It's a `ValueError`, but I don't think the changelog entry needs to include that level of detail.

File: changelog/4577.breaking.rst
Comment (ayshih) [inline changelog/4577.breaking.rst:1]:
```suggestion
The ``.size`` property of a coordinate frame with no associated data will now raise an error instead of returning 0.
```

File: sunpy/coordinates/tests/test_metaframes.py
Comment (ayshih) [inline sunpy/coordinates/tests/test_metaframes.py:122]:
```suggestion
```

Comment (ayshih) [reply sunpy/coordinates/tests/test_metaframes.py:122]: On second thought, let's put a non-SunPy frame through the exact same test for SunPy frames.  No reason not to.  Below the `rot_hpc` fixture above, add the following:
```python
# Test a non-SunPy frame
@pytest.fixture
def rot_hme():
    return (HeliocentricMeanEcliptic,
            RotatedSunFrame(lon=1*u.deg, lat=2*u.deg, distance=3*u.AU,
                            base=HeliocentricMeanEcliptic(obstime='2001-01-01', equinox='2001-01-01'),
                            duration=4*u.day))
```
And then add `"rot_hme"` to the list of fixtures before `test_class_creation()`.

**Review by ayshih — CHANGES_REQUESTED** [review]
See comments

File: sunpy/coordinates/tests/test_metaframes.py
Comment (kathivashkiv) [reply sunpy/coordinates/tests/test_metaframes.py]: I had a conflict while pushing this changes to branch. Please, have a look at my commits, whether result is acceptable..

**Review by ayshih — APPROVED** [review]
We'll see if the tests are happy, but everything looks fine to me.

Since this is not a bugfix PR, a second developer needs to approve this as well.

**Comment by ayshih** [comment]
Oh, and this should definitely be rebased/squashed before merge.

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
> Oh, and this should definitely be rebased/squashed before merge.

You can use the merge button to do so if you prefer.

**Comment by kathivashkiv** [comment]
> Oh, and this should definitely be rebased/squashed before merge.

Can you please explain me how to do this right? Here or on the IRC channel?

**Comment by nabobalis** [comment]
> > Oh, and this should definitely be rebased/squashed before merge.
> 
> Can you please explain me how to do this right? Here or on the IRC channel?
> Upd: can I use this manual?
> [docs.astropy.org/en/stable/development/workflow/development_workflow.html#rebase](https://docs.astropy.org/en/stable/development/workflow/development_workflow.html#rebase)

Yeah that should work.

**Comment by ayshih** [comment]
The rebase is slightly tricky because of the conflict.  You can give it a try to see if you can figure it out.  Nothing beats learning by experience!

**Comment by kathivashkiv** [comment]
> The rebase is slightly tricky because of the conflict. You can give it a try to see if you can figure it out. Nothing beats learning by experience!

I tried "git rebase -i" and managed to change latest commits. But I get "Merge conflict" (again) when I try to change the git history in the area where I used to have a merge conflict before. So it seems really tricky for me, I'm afraid to break everything, cause I haven't solved merge conflict before. Can we resolve this in another way? (e.g. to use the github's merge button, as was mentioned above)

**Comment by ayshih** [comment]
Okay, no worries, I'll take care of the rebase.  Don't touch this branch anymore.

**Comment by kathivashkiv** [comment]
Okay, thank you

**Comment by ayshih** [comment]
I rebased and squashed the PR into two commits.  Let's see if the tests pass

**Comment by ayshih** [comment]
Test failures are unrelated, so merging, thanks!

**Comment by kathivashkiv** [comment]
Thank you for your help!

**Pull request merged**
